source/vulkan/ht_vkswapchain.cpp

```diff
@@ -0,0 +1,271 @@
+                */
+                VkPresentModeKHR choosenPresentMode = VK_PRESENT_MODE_FIFO_KHR;
+                bool presentModeFound = false;
+
+                for ( std::vector<VkPresentModeKHR>::iterator prsnt = presentModes.begin(); prsnt != presentModes.end(); ++prsnt)
+                {
+                    if ( *prsnt == VK_PRESENT_MODE_MAILBOX_KHR )
+                    {
+                        choosenPresentMode = VK_PRESENT_MODE_MAILBOX_KHR;
+                        presentModeFound = true;
+                        break;
+                    }
+
+                    if ( *prsnt == VK_PRESENT_MODE_FIFO_KHR )
+                        presentModeFound = true;
+                }
+
+                if ( !presentModeFound )
+                {
+                    HT_ERROR_PRINTF("VKSwapChain::Initialize(): Default present mode not found.\n");
+                    return false;
+                }
+
+                VkSurfaceCapabilitiesKHR surfaceCapabilities;
+                vkGetPhysicalDeviceSurfaceCapabilitiesKHR(device, m_surface, &surfaceCapabilities); 
+
+                uint32_t imageCount = surfaceCapabilities.minImageCount + 1 <= surfaceCapabilities.maxImageCount ?
+                    surfaceCapabilities.minImageCount + 1 : surfaceCapabilities.maxImageCount;
+
+                VkExtent2D choosenExtent;
+
+                if ( surfaceCapabilities.currentExtent.width != UINT32_MAX )
+                    choosenExtent = surfaceCapabilities.currentExtent;
+                else
+                {
+                    choosenExtent.width = std::max(surfaceCapabilities.minImageExtent.width, std::min(surfaceCapabilities.maxImageExtent.width, m_width));
+                    choosenExtent.height = std::max(surfaceCapabilities.minImageExtent.height, std::min(surfaceCapabilities.maxImageExtent.height, m_height));
+                }
+
+                VkSwapchainCreateInfoKHR createSwapChainInfo;
+                createSwapChainInfo.sType = VK_STRUCTURE_TYPE_SWAPCHAIN_CREATE_INFO_KHR;
+                createSwapChainInfo.surface = m_surface;
+                createSwapChainInfo.minImageCount = imageCount;
+                createSwapChainInfo.imageColorSpace = choosenSurfaceFormat.colorSpace;
+                /* I do believe in the first version we are not goig to enable occulus, vive and others... */
+                createSwapChainInfo.imageArrayLayers = 1;
+                createSwapChainInfo.imageUsage = VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT;
+                createSwapChainInfo.preTransform = surfaceCapabilities.currentTransform;
+                createSwapChainInfo.compositeAlpha = VK_COMPOSITE_ALPHA_OPAQUE_BIT_KHR;
+                createSwapChainInfo.presentMode = choosenPresentMode;
+                createSwapChainInfo.clipped = VK_TRUE;
+                createSwapChainInfo.oldSwapchain = VK_NULL_HANDLE;
+                createSwapChainInfo.imageExtent = choosenExtent;
+
```

Comment:
There need to be a few lines added here for this to run correctly.

First off, the createInfo struct requires some initialized members. Namely
the pNext variable must be null, and there must be a flags parameter set to 0

`createSwapChainInfo.pNext = nullptr`
` createSwapChainInfo.flags = 0`

Also, there is currently no image format. There needs to be a valid image format set on the swapchain
Something like:

`createSwapChainInfo.imageFormat = VK_FORMAT_B8G8R8A8_UNORM;`